src/lib/chart/ChartActivityTracker.ts

```diff
@@ -0,0 +1,57 @@
+import { trackActivity } from '$lib/activity';
+
+const SCROLL_THRESHOLD = 50;
+
+/**
+ * Track chart activity when the user interacts with the chart.
+ * - interactions include: scolling (beyond threshold), zooming, changing periodicity
+ * - zooming causes a change in scroll value, so we don't have to detect it separately
+ * - tracking stops after first interaction is detected / tracked
+ */
+export default class ChartActivityTracker {
+	chartEngine: any;
+	initialScroll: number;
+	listeners: any[] = [];
+
+	constructor(chartEngine: any) {
+		this.chartEngine = chartEngine;
+		this.initialScroll = this.chartEngine.chart.scroll;
+		this.addEventListeners();
+	}
+
+	addEventListeners() {
+		let listener = this.chartEngine.addEventListener('scroll', () => {
+			if (this.scrolledPastThreshold()) {
+				this.track();
+			}
+		});
+		this.listeners.push(listener);
+
+		listener = this.chartEngine.addEventListener('periodicity', ({ prevPeriodicity }) => {
+			if (this.periodicityChanged(prevPeriodicity)) {
+				this.track();
+			}
+		});
+		this.listeners.push(listener);
+	}
+
+	removeEventListeners() {
+		while (this.listeners.length) {
+			this.chartEngine.removeEventListener(this.listeners.pop());
+		}
+	}
+
+	scrolledPastThreshold() {
+		return Math.abs(this.initialScroll - this.chartEngine.chart.scroll) > SCROLL_THRESHOLD;
+	}
+
+	periodicityChanged(prev) {
+		const { period, interval, timeUnit } = this.chartEngine.getPeriodicity();
+		return !(prev.periodicity === period && prev.interval === interval && prev.timeUnit === timeUnit);
+	}
+
+	track() {
+		trackActivity('chart');
+		this.removeEventListeners();
```

Comment:
Smart, we remove the listener overhead after sending the tracking data for the first time. :+1:

Reply:
My first attempt at this was to have a `tracked` boolean instance var that got set to `true` after calling `track` … then I realized it's cleaner to just remove the listeners. And yea… there may be a performance benefit as well.

It's possible to push this pattern too far. I've written functions that overwrite themselves with a `noop` when they are run as a way of ensuring they are only run once. It works… but it's admittedly a bit opaque / confusing.